### PR TITLE
POST Transit task, Dataset storage rationalisation

### PR DESCRIPTION
**Pull request by calummackervoy**
@MattTennison sorry about the massive MR :grimacing: this MR shouldn't impact the front-end

Partial implementation of #3. There's going to be lots more to do for that so I thought I should merge the part that's working before I move onto the next thing

* the file storage is split up into collections of datasets now. Previously we were using a single monolith dataset for "settlements". `AbstractFileStorageWrapper`, `DatasetItem` and `DatasetCollection` provide classes for getting file storage paths and URIs. `MUDApplication` still stores the root diretories of these things and they're still not configurable in the `web.xml` (another issue)
* Generic `DataController` endpoint. Effectively I give this any path (it's the default) and it'll get me that dataset (e.g. `/mud/world/` or `/mud/actions/tasks/mytask/#myResource`. At the moment if that dataset doesn't exist it'll create it, it'd be better if it was 404
* TaskController system copies the patterns of the ContentDescriber system
* TransitActor is a specific kind of task which you can POST to, and it will create the task for you and return `201` with the location of your dataset
* The idea is that soon my Character (or Party of characters) will enter into a Transit task, which requires me to wait for them to arrive at a destination. By GETting the task after its endTime (milliseconds since 1970 in the real world), it will return me the task as complete, and the `endState` consequences can be made (updating the world state and my character's state)
* `ResourceItem`, `Patch` and `Task` classes manage the creation and TDB commit of resources

### CODE REVIEW

**Comment by MattTennison** [comment]
Sorry, been meaning to review this - I'll take a look this evening, is that okay? :smile:

**Comment by calummackervoy** [comment]
Yeah no problem at all! No need to apologise, from my point of view at this stage of the project I'm thinking "it's great if I can get a code review, but if I can't it's no problem" - because I appreciate that I'm writing a lot of code so I think some of it will have to be merged without review

File: src/main/java/com/mackervoy/calum/mud/DatasetCollection.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * A collection (directory) of datasets. Manages the URI and file-store bindings to each particular
+ */
+public class DatasetCollection extends AbstractFileStorageWrapper {
+	
+	/**
+	 * @param subPath: the directory path of the collection *relative* the the root directory, starting and ending with a "/"
+	 * e.g. /world/mydatacollection/
+	 */
+	public DatasetCollection(String subPath) {
+		this.path = subPath;
```
Comment (MattTennison) [inline src/main/java/com/mackervoy/calum/mud/DatasetCollection.java]: It looks like the functionality of `AbstractFileStorageWrapper` relies on this.path being defined so perhaps one to lift up to its constructor? As we do the same thing in `DatasetItem` as well

File: src/main/java/com/mackervoy/calum/mud/behaviour/task/TaskActorFactory.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	 * @returns true if the object could be registered, false if another class has already registered
+	 */
+	public static boolean register(String rdfType, Class<? extends ITaskActor> actor) {
+		if(taskActors.put(rdfType, actor) == null) {
+			return true;
+		}
+		return false;
+	}
+	
+	public ITaskActor getActor(String rdfType) {
```
Comment (MattTennison) [inline src/main/java/com/mackervoy/calum/mud/behaviour/task/TaskActorFactory.java]: I think we could use an Optional here, to make clear that `getActor(...)` isn't guaranteed to return a ITaskActor

File: src/main/java/com/mackervoy/calum/mud/vocabularies/MUDLogic.java [hunk 3]
```diff
@@ -31,5 +31,6 @@ protected static final Property property( String local )
     public final static Resource Task = resource( "Task" );
     public final static Resource Transit = resource( "Transit" );
     public final static Property endState = property( "endState" );    
-
+    public final static Property inserts = property( "inserts" );
+    public final static Property deletes = property( "inserts" );
```
Comment (MattTennison) [inline src/main/java/com/mackervoy/calum/mud/vocabularies/MUDLogic.java]:
```suggestion
    public final static Property deletes = property( "deletes" );  
```

Looks like a typo

**Review by MattTennison — COMMENTED** [review]
Couple of minor code style comments, nothing major. 

Nice work, ngl some of the detail went over my head but the logic outlined in your description makes sense

**Comment by calummackervoy** [comment]
Thanks, made these changes :)

**Pull request merged**